### DISCUSSION TITLE
Need info on storage of statefile and Github integrations

### DISCUSSION THREAD

**Discussion by lechnerc77** [SAP/terraform-provider-btp#967] [Q&A]
Our team is currently in the process of automating our infrastructure on BTP and Cloud Foundry. We have made significant progress but need some clarification to ensure we are following best practices. Currently, we are running Terraform commands locally, which we have identified as potentially risky.

We have a few outstanding queries from our team:

- Where can we securely store the state file for our project? Options we are considering include AWS S3, BTP Object Store, or other suitable solutions.
- How can we integrate GitHub Actions with our project?
- We also need information regarding the "Runner." Specifically, where it should be executed—whether on a server or in a Docker container.

Any expert advice on these matters would be greatly appreciated.

Thank you for your assistance.
 
Originally posted in [issues](https://github.com/SAP/terraform-provider-btp/issues/966) by <https://github.com/Singbad>

**Comment by vipinvkmenon** [comment]
> Where can we securely store the state file for our project? Options we are considering include AWS S3, BTP Object Store, or other suitable solutions.

From the `terraform provider` perspective there isn't a difference or dependency between using any of the backends. All backends supported by Terraform officially should work. It would be better to look at the [official documentation](https://developer.hashicorp.com/terraform/language/backend) from Terraform and documentation for specific backends for the best practices. For E.G [S3](https://docs.aws.amazon.com/prescriptive-guidance/latest/terraform-aws-provider-best-practices/backend.html).

We do have documented some [examples](https://github.com/SAP-samples/btp-terraform-samples/tree/main/released/terraform-be) as a reference to how backends (in general) in terraform can be used.

> How can we integrate GitHub Actions with our project?

A good source for the documentation for this would be from the [Terraform official documentation](https://developer.hashicorp.com/terraform/tutorials/automation/github-actions) as this is also generic to Terraform and not the providers themselves. 
While we don't have any explicit sample for this, you could refer to the way we have setup our terraform [test pipeline](https://github.com/SAP-samples/btp-terraform-samples/blob/main/.github/workflows/scenario-test.yml).

> We also need information regarding the "Runner." Specifically, where it should be executed—whether on a server or in a Docker container.

What do you mean by server or docker? In all cases, the github actions would run within containers. Could you elaborate on the question?  Do you mean `self-hosted` servers vs `github-hosted` servers?

**Reply by Singbad** [reply]
Github-Hosted Servers maybe?

**Reply by lechnerc77** [reply]
See answer below on the runners

**Reply by amanikalo** [reply]
Hello colleagues,
Thanks for your replies. 
Regarding the statefile storage. Does storing it in S3 goes well with "SAP runs SAP" principle? Is it the official recommendation from Terraform Provider for BTP team? Probably @lechnerc77 can comment on this

**Reply by lechnerc77** [reply]
We do not have a recommendation concerning the storage of the state file. You can and should use what works best in your organization and what is supported by Terraform.

We do not offer a SAP-managed state backend, so the choices are defined by what is supported by Terraform.

**Comment by lechnerc77** [comment]
As @vipinvkmenon was faster: Concerning the runner of GitHub Actions: That depends on your setup. I assume all necessayr information about runners and the pros and cons are available here: https://docs.github.com/en/actions/using-github-hosted-runners/using-github-hosted-runners

**Comment by lechnerc77** [comment]
Concerning examples for GH Action workflows: You also find some in the org https://github.com/btp-automation-scenarios 

Keep in mind these are **samples** and are not applicable 1:1 in productive setups
